### PR TITLE
[JENKINS-54391] Hook to verify execution logs

### PR DESCRIPTION
**Pull request by varyvol**
[JENKINS-54391](https://issues.jenkins-ci.org/browse/JENKINS-54391)

Create hook to verify the logs from the execution of a Jenkinsfile within a JFR image. Also adapt smoke tests to use such hook.

### CODE REVIEW

File: src/hooks/logs.inc [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+set -e
+
+# Check if the execution log of some command contains an expected pattern
+# $1: Expected contained value
+# $2: (Optional) Logs from the test or returned message
+logs_contains() {
+
+    if [ "$#" -ge 1 ]
+    then
+        if [ -z "${_shunit_test_}" ]
```
Comment (fcojfernandez) [inline src/hooks/logs.inc]: This piece of code can be refactored into a function. I did something similar in https://github.com/jenkinsci/jenkinsfile-runner-test-framework/blob/03695f90ac9dca39976075b45a3ecefa54eadb3b/src/hooks/workspace.inc#L7

File: src/hooks/logs.inc [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            assertEquals "$test_name should contain '$1'" "1" "$text_contained"
+            return 0
+        else
+            echo "Error. No log to inspect"
+            return 1
+        fi
+    else
+        echo "Error. Missing parameters:"
+        echo "   Expected contained value"
+        echo "   (Optional) Returned message"
```
Comment (fcojfernandez) [inline src/hooks/logs.inc]: Not sure this message is correct

File: src/hooks/logs.inc [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            assertEquals "$test_name should not contain '$1'" "0" "$text_contained"
+            return 0
+        else
+            echo "Error. No log to inspect"
+            return 1
+        fi
+    else
+        echo "Error. Missing parameters:"
+        echo "   Expected contained value"
+        echo "   (Optional) Returned message"
```
Comment (fcojfernandez) [inline src/hooks/logs.inc]: ditto

File: src/hooks/result.inc [hunk 4]
```diff
@@ -1,6 +1,11 @@
 #!/bin/bash
 set -e
 
+# Path to this source file, stored in BASH_SOURCE
+root_directory=$(dirname "${BASH_SOURCE[0]}")
+# shellcheck source=src/utilities/utils.inc
+. "$root_directory"/logs.inc
```
Comment (fcojfernandez) [inline src/hooks/result.inc]: Maybe include this _inc_ file in the `init-jfr-test-framwork.inc` script?

File: src/hooks/result.inc [hunk 5]
```diff
@@ -25,66 +30,20 @@ execution_success() {
...
-    fi
-
-}
-
 # Check a generic execution is ok
 # $1: Returned code by the process
-# $2: Execution log
-# $3: Expected excerpt
+# $2: Expected excerpt
+# $2: (Optional) Execution log
```
Comment (fcojfernandez) [inline src/hooks/result.inc]: $3
And I think you should check and update the else block: https://github.com/jenkinsci/jenkinsfile-runner-test-framework/pull/10/files#diff-7be02b3ea4ffd6436873700536acd999R47

File: src/hooks/result.inc [hunk 6]
```diff
@@ -25,66 +30,20 @@ execution_success() {
...
 execution_should_success() {
 
     if [ "$#" -eq 3 ]
     then
         execution_success "$1"
-        message_contains "$2" "$3"
+        logs_contains "$2" "$3"
+    elif [ "$#" -eq 3 ]
```
Comment (fcojfernandez) [inline src/hooks/result.inc]: I think ti should be 2
```suggestion
    elif [ "$#" -eq 2 ]
```

File: src/hooks/result.inc [hunk 7]
```diff
@@ -97,14 +56,19 @@ execution_should_success() {
 
 # Check the execution of the Jenkinsfile should succed
 # $1: Returned code by the process
-# $2: Execution log
+# $2: (Optional) Execution log
```
Comment (fcojfernandez) [inline src/hooks/result.inc:54]: Since it's Optional, you should update the `echo` command in the else block

File: src/hooks/result.inc [hunk 8]
```diff
@@ -97,14 +56,19 @@ execution_should_success() {
...
 
     if [ "$#" -eq 2 ]
     then
         execution_success "$1"
-        message_contains "$2" "[Pipeline] End of Pipeline"
-        message_contains "$2" "Finished: SUCCESS"
+        logs_contains "[Pipeline] End of Pipeline" "$2"
+        logs_contains "Finished: SUCCESS" "$2"
+    elif [ "$#" -eq 1 ]
+        then
```
Comment (fcojfernandez) [inline src/hooks/result.inc]: indentation

File: src/hooks/result.inc [hunk 9]
```diff
@@ -116,15 +80,21 @@ jenkinsfile_execution_should_succed() {
 
 # Check the execution of the Jenkinsfile should fail
 # $1: Returned code by the process
-# $2: Execution log
+# $2: (Optional) Execution log
```
Comment (fcojfernandez) [inline src/hooks/result.inc:78]: ditto

File: src/hooks/result.inc [hunk 10]
```diff
@@ -136,15 +106,21 @@ jenkinsfile_execution_should_fail() {
 
 # Check the execution of the Jenkinsfile should be unstable
 # $1: Returned code by the process
-# $2: Execution log
+# $2: (Optional) Execution log
```
Comment (fcojfernandez) [inline src/hooks/result.inc:104]: ditto

File: src/jfr/jenkinsfile-runner.inc [hunk 11]
```diff
@@ -11,9 +11,19 @@ run_jfr_docker_image() {
     then
         if [ -z "$JAVA_OPTS" ]
         then
-            docker run --rm -v "$2":/workspace/Jenkinsfile "$1"
+            if [ ! -z "${_shunit_test_}" ]
+            then
+                docker run --rm -v "$2":/workspace/Jenkinsfile "$1" 2>&1 | tee "${_shunit_test_}.log"
```
Comment (fcojfernandez) [inline src/jfr/jenkinsfile-runner.inc:16]: nice!

**Review by fcojfernandez — COMMENTED** [review]
@varyvol a couple of questions:

1. Have you check the syntax? (Using `make syntax`)
1. You're adding elif blocks to pass the optional parameter. Wouldn't it work if you keep only the if block and "$2" is ""?

File: src/hooks/logs.inc (hunk 2 above)
Comment (varyvol) [reply src/hooks/logs.inc]: Changed.

File: src/hooks/logs.inc (hunk 3 above)
Comment (varyvol) [reply src/hooks/logs.inc]: Changed.

File: src/hooks/result.inc (hunk 4 above)
Comment (varyvol) [reply src/hooks/result.inc]: It's done (look at the first file changed in this PR).

File: src/hooks/result.inc (hunk 5 above)
Comment (varyvol) [reply src/hooks/result.inc]: Changed.

File: src/hooks/result.inc (hunk 6 above)
Comment (varyvol) [reply src/hooks/result.inc]: Changed.

File: src/hooks/result.inc (hunk 7 above)
Comment (varyvol) [reply src/hooks/result.inc:54]: Changed.

File: src/hooks/result.inc (hunk 8 above)
Comment (varyvol) [reply src/hooks/result.inc]: Changed.

File: src/hooks/result.inc (hunk 9 above)
Comment (varyvol) [reply src/hooks/result.inc:78]: Changed.

File: src/hooks/result.inc (hunk 10 above)
Comment (varyvol) [reply src/hooks/result.inc:104]: Changed.

File: src/hooks/logs.inc (hunk 1 above)
Comment (varyvol) [reply src/hooks/logs.inc]: Refactored.

**Comment by varyvol** [comment]
@fcojfernandez yes, I had checked the syntax (and have done so again). As per your second point, to be honest I did not test it but I find the code to be more readable and understandable like that.

File: src/hooks/result.inc (hunk 4 above)
Comment (fcojfernandez) [reply src/hooks/result.inc]: Since it's intended to load the framework using that script, it's not needed to load here. You will be loading twice

**Review by fcojfernandez — APPROVED** [review]
Just the nit about the load of the script

File: src/hooks/result.inc (hunk 4 above)
Comment (varyvol) [reply src/hooks/result.inc]: Removed then!

**Pull request merged**